ProcessMaker/Jobs/DownloadSecurityLog.php

```diff
@@ -37,18 +50,161 @@ public function __construct(User $user, string $format, int $userId = null)
+            $expires = $this->getExpires();
+            $filename = $this->createTemporaryFilename();
+            $url = $this->createTemporaryUrl($filename, $contents, $expires);
+            $message = __('Click on the link and download the file. This link will be available until midnight tonight.');
+            event(new SecurityLogDownloadJobCompleted($this->user, true, $message, $url));
+        } catch (Exception $e) {
+            $message = __('Sorry, it was not possible to generate the log file. Please contact the administrator.');
+            event(new SecurityLogDownloadJobCompleted($this->user, false, $message));
+        }
+    }
+
+    /**
+     * Generate the content with the corresponding format
+     */
+    protected function generateContent(array $collection)
+    {
+        return $this->format === static::FORMAT_CSV ? $this->toCSV($collection) : $this->toXML($collection);
+    }
+
+    /**
+     * To CSV
+     */
+    protected function toCSV(array $collection)
+    {
+        $first = current($collection);
+        $headers = array_keys($first ?: []);
+        $content = '';
+        $line = [];
+        foreach ($headers as $key) {
+            $line[] = $key;
+        }
+        $content .= implode(static::CSV_SEPARATOR, $line);
+        foreach ($collection as $item) {
+            $line = [];
+            foreach ($headers as $key) {
+                if (is_object($item[$key])) {
+                    $line[] = json_encode($item[$key]);
+                } else {
+                    $line[] = isset($item[$key]) ? $item[$key] : null;
+                }
+            }
+            $content .= PHP_EOL . implode(static::CSV_SEPARATOR, $line);
         }
+        return $content;
+    }
+
+    /**
+     * To XML
+     */
+    protected function toXML(array $collection)
+    {
+        $content = '<?xml version="1.0" encoding="UTF-8"?>' . PHP_EOL;
+        $content .= '<securityLogs>';
+        $tab = "\t";
+        foreach ($collection as $item) {
+            $content .= PHP_EOL . $tab . '<securityLog>';
+            foreach ($item as $key => $value) {
+                if (is_object($value)) {
+                    $value = json_encode($value);
+                }
+                $content .= sprintf(
+                    '%s<%s>%s</%s>',
+                    PHP_EOL . $tab,
+                    $key,
+                    $value,
+                    $key
+                );
+            }
+            $content .= PHP_EOL . $tab . '</securityLog>';
+        }
+        $content .= PHP_EOL . '</securityLogs>';
+
+        return $content;
+    }
+
+    /**
+     * Get collection
+     *
+     * @return array
+     */
+    protected function getCollection()
+    {
+        $query = SecurityLog::query();
+        if ($this->userId) {
+            $query->where('user_id', $this->userId);
+        }
+        $results = [];
+        $query->chunk(1000, function (Collection $collection) use (&$results) {
```

Comment:
You are chunking the result by 1000, but inside this function you are placing all the content in an array which could use large amount of memory. Consider to use `cursor()` to prepare and stream the final result to the S3 file.